File: Controllers/ProductController.vb
Imports System.Web.Mvc
Imports My.Models

Namespace My.Controllers
    Public Class ProductController
        Inherits Controller

        ' GET: /Product
        Function Index(categoryId As Integer?, keyword As String, page As Integer?) As ActionResult
            Dim viewModel As New ProductListViewModel With {
                .CurrentCategoryId = categoryId,
                .SearchKeyword = keyword,
                .PageIndex = If(page, 1),
                .PageSize = 12
            }
            ' TODO: 從資料庫查詢商品列表
            Return View(viewModel)
        End Function

        ' GET: /Product/Detail/5
        Function Detail(id As Integer) As ActionResult
            ' TODO: 從資料庫查詢商品詳細資料
            Dim product As Product = Nothing
            If product Is Nothing Then Return HttpNotFound()
            Return View(product)
        End Function

        ' GET: /Product/Category/coffee
        Function Category(slug As String) As ActionResult
            ' TODO: 依分類顯示商品
            Return View()
        End Function

    End Class
End Namespace

File: Controllers/ProductController.cs
using System.Linq;
using System.Web.Mvc;
using MyWeb.Models;

namespace MyWeb.Controllers
{
    public class ProductController : Controller
    {
        // GET: /Product
        public ActionResult Index(int? categoryId, string keyword, int? page)
        {
            int pageIndex = page ?? 1;
            const int pageSize = 12;

            using (var db = new AppDbContext())
            {
                var query = db.Products.Include("Category").Where(p => p.IsActive);

                if (categoryId.HasValue)
                    query = query.Where(p => p.CategoryId == categoryId.Value);

                if (!string.IsNullOrWhiteSpace(keyword))
                    query = query.Where(p => p.Name.Contains(keyword) || p.Description.Contains(keyword));

                int total = query.Count();
                var products = query
                    .OrderBy(p => p.SortOrder)
                    .ThenByDescending(p => p.CreatedAt)
                    .Skip((pageIndex - 1) * pageSize)
                    .Take(pageSize)
                    .ToList();

                var vm = new ProductListViewModel
                {
                    Products = products,
                    Categories = db.ProductCategories.Where(c => c.IsActive).OrderBy(c => c.SortOrder).ToList(),
                    CurrentCategoryId = categoryId,
                    SearchKeyword = keyword,
                    TotalCount = total,
                    PageIndex = pageIndex,
                    PageSize = pageSize
                };
                return View(vm);
            }
        }

        // GET: /Product/Detail/5
        public ActionResult Detail(int id)
        {
            using (var db = new AppDbContext())
            {
                var product = db.Products.Include("Category").FirstOrDefault(p => p.Id == id && p.IsActive);
                if (product == null) return HttpNotFound();
                return View(product);
            }
        }
    }
}
